test/truefi2/CrvBaseRateOracle.test.ts

```diff
@@ -0,0 +1,251 @@
+    it('correctly sets fields', async () => {
+      expect(await crvBaseRateOracle.curve()).to.eq(mockCurve.address)
+      expect(await crvBaseRateOracle.cooldownTime()).to.eq(COOLDOWN_TIME)
+    })
+
+    it('fills up one field in historical buffer', async () => {
+      const [baseRates, timestamps] = await crvBaseRateOracle.getHistBuffer()
+      expect(baseRates[0]).to.eq(0)
+      expect(timestamps[0]).to.eq(0)
+      expect(baseRates[BUFFER_SIZE - 1]).to.eq(STARTING_RATE)
+      expect(timestamps[BUFFER_SIZE - 1]).to.eq(DEPLOYMENT_TIMESTAMP)
+    })
+  })
+
+  describe('HistoricalRatesBuffer', () => {
+    it('has expected capacity', async () => {
+      const [baseRates, timestamps] = await crvBaseRateOracle.getHistBuffer()
+      expect(baseRates.length).to.eq(MAX_BUFFER_SIZE)
+      expect(timestamps.length).to.eq(MAX_BUFFER_SIZE)
+    })
+
+    it('has expected initial insert index', async () => {
+      const [, , insertIndex] = await crvBaseRateOracle.getHistBuffer()
+      expect(insertIndex).to.eq(0)
+    })
+
+    it('insertIndex increments cyclically', async () => {
+      await mockCurve.mock.get_virtual_price.returns(100)
+      for (let i = 0; i < (BUFFER_SIZE - 1); i++) {
+        await updateRateRightAfterCooldown()
+      }
+      let [, , insertIndex] = await crvBaseRateOracle.getHistBuffer()
+      expect(insertIndex).to.eq(BUFFER_SIZE - 1)
+      await updateRateRightAfterCooldown()
+      ;[, , insertIndex] = await crvBaseRateOracle.getHistBuffer()
+      expect(insertIndex).to.eq(0)
+    })
+
+    it('overwrites old values with new ones', async () => {
+      await mockCurve.mock.get_virtual_price.returns(100)
+      for (let i = 0; i < BUFFER_SIZE; i++) {
+        await updateRateRightAfterCooldown()
+      }
+      let [baseRates] = await crvBaseRateOracle.getHistBuffer()
+      expect(baseRates[0]).to.eq(100)
+      await mockCurve.mock.get_virtual_price.returns(200)
+      await updateRateRightAfterCooldown()
+      ;[baseRates] = await crvBaseRateOracle.getHistBuffer()
+      expect(baseRates[0]).to.eq(200)
+    })
+  })
+
+  describe('updateRate', () => {
+    it('reverts if cooldown is on', async () => {
+      await updateRateRightAfterCooldown()
+      await expect(crvBaseRateOracle.updateRate())
+        .to.be.revertedWith('CrvBaseRateOracle: Buffer on cooldown')
+      await timeTravel(provider, COOLDOWN_TIME)
+      await expect(crvBaseRateOracle.updateRate())
+        .not.to.be.reverted
+    })
+
+    it('adds one rate to buffer', async () => {
+      await mockCurve.mock.get_virtual_price.returns(100)
+      await updateRateRightAfterCooldown()
+      const curTimestamp = await getCurrentTimestamp()
+      const [baseRates, timestamps, insertIndex] = await crvBaseRateOracle.getHistBuffer()
+      expect(baseRates[0]).to.eq(100)
+      expect(timestamps[0]).to.eq(curTimestamp)
+      expect(insertIndex).to.eq(1)
+    })
+  })
+
+  describe('calculateAverageRate', () => {
```

Comment:
Could you write in every comment what the expected value is roughly equal to to see if it make sense?